cpp/src/arrow/compute/expression.cc

```diff
@@ -1148,6 +1149,34 @@ Result<Expression> Canonicalize(Expression expr, compute::ExecContext* exec_cont
 
 namespace {
 
+/// Preprocess an `is_in` predicate value set for simplification.
+/// \pre `value_set` is non-empty
+/// \return the value set sorted with duplicate and null values removed
+Result<std::shared_ptr<Array>> PrepareIsInValueSet(std::shared_ptr<Array> value_set) {
+  DCHECK_GT(value_set->length(), 0);
+  ARROW_ASSIGN_OR_RAISE(value_set, Unique(value_set));
+  ARROW_ASSIGN_OR_RAISE(
+      std::shared_ptr<Array> sort_indices,
+      SortIndices(value_set, SortOptions({}, NullPlacement::AtEnd)));
+  ARROW_ASSIGN_OR_RAISE(
+      value_set,
+      Take(*value_set, *sort_indices, TakeOptions(/*bounds_check=*/false)));
+  if (value_set->IsNull(value_set->length() - 1)) {
+    value_set = value_set->Slice(0, value_set->length() - 1);
+  }
+  return value_set;
+}
+
+/// Context for expression simplification.
+struct SimplificationContext {
+  /// Mapping from `is_in` calls to simplified value sets.
+  ///
+  /// `is_in` predicates with large value sets can be expensive to bind, so we
+  /// accumulate simplifications in the context and defer binding until all
+  /// inequalities have been processed.
+  std::unordered_map<const Expression::Call*, std::shared_ptr<Array>> is_in_value_sets;
```

Comment:
Might be better to add a new nullable field to `Expression::Call` called `value_set_for_is_in`.

Reply:
And using pointers as hash table keys is not a good idea.